test/system_reference_blas.hpp

```diff
@@ -26,56 +26,140 @@
-
-#undef ENABLE_SYSTEM_GEMV
-
-#define ENABLE_SYSTEM_GER(_type, _system_name)                            \
-  namespace reference_blas {                                              \
-  extern "C" void _system_name(const int *, const int *, const _type *,   \
-                               const _type *, const int *, const _type *, \
-                               const int *, _type *, const int *);        \
-  void ger(int m, int n, _type alpha, const _type a[], int incX,          \
-           const _type b[], int incY, _type c[], int lda) {               \
-    _system_name(&m, &n, &alpha, a, &incX, b, &incY, c, &lda);            \
-  }                                                                       \
-  }  // namespace reference_blas
-
-ENABLE_SYSTEM_GER(float, sger_)
-ENABLE_SYSTEM_GER(double, dger_)
-
-#undef ENABLE_SYSTEM_GER
-#define ENABLE_SYSTEM_GEMM(_type, _system_name)                               \
-  namespace reference_blas {                                                  \
-  extern "C" void _system_name(const char *, const char *, const int *,       \
-                               const int *, const int *, const _type *,       \
-                               const _type *, const int *, const _type *,     \
-                               const int *, const _type *, _type *,           \
-                               const int *);                                  \
-  void gemm(const char *transA, const char *transB, int m, int n, int k,      \
-            _type alpha, const _type a[], int lda, const _type b[], int ldb,  \
-            _type beta, _type c[], int ldc) {                                 \
-    _system_name(transA, transB, &m, &n, &k, &alpha, a, &lda, b, &ldb, &beta, \
-                 c, &ldc);                                                    \
-  }                                                                           \
-  }  // namespace reference_blas
-
-ENABLE_SYSTEM_GEMM(float, sgemm_)
-ENABLE_SYSTEM_GEMM(double, dgemm_)
-
-#undef ENABLE_SYSTEM_GEMM
+#include <iostream>
+
+#include "cblas.h"
+
+namespace {
+CBLAS_TRANSPOSE c_trans(char x) {
+  switch (x) {
+    case 't':
+    case 'T':
+      return CblasTrans;
+    case 'n':
+    case 'N':
+      return CblasNoTrans;
+    case 'c':
+    case 'C':
+      return CblasConjTrans;
+    default:
+      std::cerr << "Transpose value " << x << " is invalid.\n";
+      abort();
+  }
+}
+
+CBLAS_UPLO c_uplo(char x) {
+  switch (x) {
+    case 'u':
+    case 'U':
+      return CblasUpper;
+    case 'l':
+    case 'L':
+      return CblasLower;
+    default:
+      std::cerr << "Uplo value " << x << " is invalid.\n";
+      abort();
+  }
+}
+
+CBLAS_DIAG c_diag(char x) {
+  switch (x) {
+    case 'u':
+    case 'U':
+      return CblasUnit;
+    case 'n':
+    case 'N':
+      return CblasNonUnit;
+  }
+}
+}  // namespace
+
+namespace reference_blas {
+
+template <typename scalar_t, typename float_t, typename double_t>
+struct f_or_d {};
```

Comment:
You can use c++ variadic template with function pointer and its parameters as a template function input to call:
template<typename func, typename ... params> func(func f, params... prs) {
f(pr...);
}

Reply:
I'm not that familiar with varadic templates, I'm afraid. But is there a way to get the nth parameter of an argument and template bundle? There seems to be a way using std::tuple?

I'm not what one would gain from using varadic templates in this case either.

Reply:
it is more for maintainability. As the next step is adding conjugate version which requires two more types to be added

Reply:
I've been staring at this a while now, and I can't for the life of me figure out how I would use variadic templates for this.

Reply:
It can be something like this
template<typename type, typename func_t1, typename funct_2> struct Func;
template<typename FuncT1, Funct2> struct Func<float, FunctT1, FuncT2> struct Func{

template <typename Ret, typename ... Args> Ret cal_func(Args .... args) {
return FuncT1(args...); 
}
};
// similarly for double 
in your code you call Func<datatype, sxxx, dxxx>::calfunc(parameters);